crates/core/src/subscription/subscription.rs

```diff
@@ -262,178 +314,244 @@ impl JoinSide {
-            probe_side,
-        }))
+        Some((index_side, probe_side))
     }
 
     /// Evaluate this [`IncrementalJoin`].
-    ///
-    /// The following assumptions are made for the incremental evaluation to be
-    /// correct without maintaining a materialized view:
-    ///
-    /// * The join is a primary foreign key semijoin, i.e. one row from the
-    ///   right table joins with at most one row from the left table.
-    /// * The rows in the [`DatabaseTableUpdate`]s on either side of the join
-    ///   are already committed to the underlying "physical" tables.
-    /// * We maintain set semantics, i.e. no two rows with the same value can appear in the result.
-    ///
-    /// Based on this, we evaluate the join as:
-    ///
-    /// ```text
-    ///     let inserts = {A+ join B} U {A join B+}
-    ///     let deletes = {A- join B} U {A join B-} U {A- join B-}
-    ///
-    ///     (deletes \ inserts) || (inserts \ deletes)
-    /// ```
-    ///
-    /// Where:
-    ///
-    /// * `A`:  Committed table to the LHS of the join.
-    /// * `B`:  Committed table to the RHS of the join.
-    /// * `+`:  Virtual table of only the insert operations against the annotated table.
-    /// * `-`:  Virtual table of only the delete operations against the annotated table.
-    /// * `U`:  Set union.
-    /// * `\`:  Set difference.
-    /// * `||`: Concatenation.
-    ///
-    /// For a more in-depth discussion, see the [module-level documentation](./index.html).
-    pub fn eval(&self, db: &RelationalDB, tx: &Tx, auth: &AuthCtx) -> Result<impl Iterator<Item = TableOp>, DBError> {
-        let mut inserts = {
-            // Replan query after replacing the indexed table with a virtual table,
-            // since join order may need to be reversed.
-            let (join_a, join_a_sources) = with_delta_table(self.join.clone(), Some(self.index_side.inserts()), None);
-            let join_a = QueryExpr::from(join_a).optimize(&|table_id, table_name| db.row_count(table_id, table_name));
-
-            // No need to replan after replacing the probe side with a virtual table,
-            // since no new constraints have been added.
-            let (join_b, join_b_sources) = with_delta_table(self.join.clone(), None, Some(self.probe_side.inserts()));
-            let join_b = join_b.into();
 
+    pub fn eval<'a>(
+        &self,
+        db: &RelationalDB,
+        tx: &Tx,
+        updates: impl IntoIterator<Item = &'a DatabaseTableUpdate>,
+    ) -> Result<Option<impl Iterator<Item = TableOp>>, DBError> {
+        let Some((index_side, probe_side)) = self.find_updates(updates) else {
+            return Ok(None);
+        };
+        let mut inserts = {
             // {A+ join B}
-            let a = eval_updates(db, *auth, tx, &join_a, self.join.return_index_rows, join_a_sources)?;
+            let index_inserts = index_side.inserts();
+            let index_inserts = to_mem_table_with_op_type(
+                self.index_table.head.clone(),
```

Comment:
Optional: We can further optimize this since we know this is the header of a MemTable, and so we can also compute the Header ahead of time.